pilot/pkg/model/network.go

```diff
@@ -142,11 +152,19 @@ func (mgr *NetworkManager) reload() NetworkGatewaySet {
 	// - the internal map of label gateways - these get deleted if the service is deleted, updated if the ip changes etc.
 	// - the computed map from meshNetworks (triggered by reloadNetworkLookup, the ported logic from getGatewayAddresses)
 	gatewaySet.InsertAll(mgr.env.NetworkGateways()...)
-	mgr.multiNetworkEnabled = len(gatewaySet) > 0
+	resolvedGatewaySet := mgr.resolveHostnameGateways(gatewaySet)
 
-	mgr.resolveHostnameGateways(gatewaySet)
+	return mgr.NetworkGateways.update(resolvedGatewaySet) || mgr.Unresolved.update(gatewaySet)
```

Comment:
I am not sure this is safe wrt to mutex usage anymore. We can have 2 concurrent calls to reload(), and have the results mixed up ?

Reply:
Good point. Would you say 3 mutexes is safe? 

2 in the NetworkGateways structs for read protection and then 1 in the NetworkManager to ensure things are updated together? 

A bit awkward either way. Could also maybe share a reference to the "root" mutex but that feels even more awkward.